File: src/test/java/kitchenpos/product/tobe/application/ProductServiceTest.java
package kitchenpos.product.tobe.application;


import kitchenpos.common.tobe.Profanities;
import kitchenpos.menu.tobe.fake.InMemoryMenuRepository;
import kitchenpos.product.tobe.domain.*;
import kitchenpos.product.tobe.fake.FakePurogmalumClient;
import kitchenpos.product.tobe.fake.InMemoryProductRepository;
import kitchenpos.product.tobe.fixture.ProductFixture;
import kitchenpos.menu.tobe.domain.menu.MenuRepository;

import kitchenpos.product.tobe.infra.ProductPriceValidator;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.context.ApplicationEventPublisher;

import java.util.NoSuchElementException;
import java.util.UUID;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;

class ProductServiceTest {
    private ProductRepository productRepository;
    private ProductService productService;
    private ProductValidator productValidator;
    private Profanities profanities;
    private ApplicationEventPublisher applicationEventPublisher;

    @BeforeEach
    void setUp() {
        productRepository = new InMemoryProductRepository();
        MenuRepository menuRepository = new InMemoryMenuRepository();
        profanities = new FakePurogmalumClient("바보");
        productValidator = new ProductPriceValidator(menuRepository);
        productService = new ProductService(productRepository, profanities, applicationEventPublisher);
    }

    @Nested
    @DisplayName("상품 생성")
    class CreateProduct {
        @Test
        @DisplayName("성공")
        void success() {
            Product request = ProductFixture.product("후라이드", 16000, profanities);

            Product created = productService.create(request);
            assertAll(
                    () -> assertThat(created.getId()).isNotNull(),
                    () -> assertThat(created.getName()).isEqualTo(new ProductName("후라이드", profanities)),
                    () -> assertThat(created.getProductPrice()).isEqualTo(new ProductPrice(16000L))
            );
        }

        @Test
        @DisplayName("음수 가격으로 생성 실패")
        void failWithNegativePrice() {

            assertThatThrownBy(() -> productService.create(ProductFixture.product("후라이드", -1000, profanities)))
                    .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        @DisplayName("비속어 포함된 이름으로 생성 실패")
        void failWithProfanity() {

            assertThatThrownBy(() -> productService.create(ProductFixture.product("바보", 1000, profanities)))
                    .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    @DisplayName("가격 변경")
    class ChangePrice {
        @Test
        @DisplayName("성공")
        void success() {
            Product product = ProductFixture.product("후라이드", 16000, profanities);
            productRepository.save(product);
            Product request = ProductFixture.product("후라이드", 18000, profanities);

            Product updated = productService.changePrice(product.getId(), request);

            assertThat(updated.getProductPrice().equals(new ProductPrice(18000L)));
        }

        @Test
        @DisplayName("존재하지 않는 상품 실패")
        void failWithNonExistentProduct() {
            UUID nonExistentId = UUID.randomUUID();
            Product request = ProductFixture.product("후라이드", 1000, profanities);

            assertThatThrownBy(() -> productService.changePrice(nonExistentId, request))
                    .isInstanceOf(NoSuchElementException.class);
        }

        @Test
        @DisplayName("음수 가격 변경 실패")
        void failWithNegativePrice() {
            Product product = ProductFixture.product("후라이드", 16000, profanities);
            productRepository.save(product);

            assertThatThrownBy(() -> productService.changePrice(product.getId(), ProductFixture.product("후라이드", -1000, profanities)))
                    .isInstanceOf(IllegalArgumentException.class);
        }
    }
}

File: src/test/kotlin/kitchenpos/product/tobe/application/ProductServiceTest.kt
package kitchenpos.product.tobe.application

import java.math.BigDecimal
import kitchenpos.common.domain.Profanities
import kitchenpos.menu.tobe.domain.MenuProductPriceChanged
import kitchenpos.menu.tobe.domain.MenuRepository
import kitchenpos.menu.tobe.domain.ProductClient
import kitchenpos.menu.tobe.infra.DefaultMenuAmountService
import kitchenpos.menu.tobe.infra.DefaultMenuProductPriceChanged
import kitchenpos.menu.tobe.infra.DefaultProductClient
import kitchenpos.menu.tobe.infra.FakeMenuRepository
import kitchenpos.product.tobe.application.dto.ChangeProductPriceReq
import kitchenpos.product.tobe.application.dto.CreateProductReq
import kitchenpos.product.tobe.domain.ProductRepository
import kitchenpos.product.tobe.infra.FakeProductRepository
import kitchenpos.product.tobe.infra.FakeProfanities
import kitchenpos.utils.Fixtures
import org.assertj.core.api.Assertions
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

class ProductServiceTest {
    private lateinit var productRepository: ProductRepository
    private lateinit var productClient: ProductClient
    private lateinit var menuRepository: MenuRepository
    private lateinit var profanities: Profanities
    private lateinit var productService: ProductService
    private lateinit var menuProductPriceChanged: MenuProductPriceChanged

    @BeforeEach
    fun setUp() {
        productRepository = FakeProductRepository()
        productClient = DefaultProductClient(productRepository)
        menuRepository = FakeMenuRepository()
        profanities = FakeProfanities()
        menuProductPriceChanged =
            DefaultMenuProductPriceChanged(menuRepository, DefaultMenuAmountService(productClient = productClient))
        productService = ProductService(productRepository, profanities, menuProductPriceChanged)
    }

    @Test
    @DisplayName("`Product`를 등록한다")
    fun createProduct() {
        // given
        val request = CreateProductReq("양념치킨", BigDecimal.valueOf(16000))

        // when
        val product = productService.create(request)

        // then
        assertThat(product.name).isEqualTo("양념치킨")
        assertThat(product.price).isEqualTo(BigDecimal.valueOf(16000))
    }

    @Test
    @DisplayName("`Product`를 전체 조회한다")
    fun findAll() {
        // given
        val request1 = CreateProductReq("양념치킨", BigDecimal.valueOf(16000))
        val request2 = CreateProductReq("후라이드치킨", BigDecimal.valueOf(17000))
        productService.create(request1)
        productService.create(request2)

        // when
        val products = productService.findAll()

        // then
        assertThat(products.size).isEqualTo(2)
    }

    @Test
    @DisplayName("`Product`의 `price`를 변경한다")
    fun changePrice() {
        // given
        val product = productRepository.save(Fixtures.product(name = "양념치킨", price = 16000))

        // when
        productService.changePrice(product.id, ChangeProductPriceReq(BigDecimal.valueOf(17000)))

        // then
        assertThat(product.productPrice.price).isEqualTo(BigDecimal.valueOf(17000))
    }

    @Test
    @DisplayName("등록되지 않는 `Product`의 `price`는 변경할 수 없다")
    fun changePriceFail() {
        // when then
        Assertions.assertThatThrownBy {
            productService.changePrice(Fixtures.INVALID_UUID, ChangeProductPriceReq(BigDecimal.valueOf(17000)))
        }.isInstanceOf(NoSuchElementException::class.java)
    }
}
